component/common/src/main/java/org/wso2/carbon/extension/identity/emailotp/common/EmailOtpServiceImpl.java

```diff
@@ -0,0 +1,351 @@
+            }
+            return null;
+        }
+        SessionDTO previousOTPSessionDTO;
+        try {
+            previousOTPSessionDTO = new ObjectMapper().readValue(jsonString, SessionDTO.class);
+        } catch (IOException e) {
+            throw Utils.handleServerException(Constants.ErrorMessage.SERVER_JSON_SESSION_MAPPER_ERROR, null, e);
+        }
+        // If the previous OTP is issued within the interval, return the same.
+        int otpRenewalInterval = EmailOtpServiceDataHolder.getConfigs().getOtpRenewalInterval();
+        long elapsedTime = System.currentTimeMillis() - previousOTPSessionDTO.getGeneratedTime();
+        boolean isValidToResend = elapsedTime < otpRenewalInterval;
+        if (isValidToResend) {
+            return previousOTPSessionDTO;
+        }
+        return null;
+    }
+
+    private void persistOTPSession(SessionDTO sessionDTO, String sessionId) throws EmailOtpServerException {
+
+        String jsonString;
+        try {
+            jsonString = new ObjectMapper().writeValueAsString(sessionDTO);
+        } catch (JsonProcessingException e) {
+            throw Utils.handleServerException(
+                    Constants.ErrorMessage.SERVER_SESSION_JSON_MAPPER_ERROR, e.getMessage(), e);
+        }
+        SessionDataStore.getInstance().storeSessionData(sessionId, Constants.SESSION_TYPE_OTP, jsonString,
+                getTenantId());
+        if (log.isDebugEnabled()) {
+            log.debug(String.format("Successfully persisted the OTP for the user Id: %s.", sessionDTO.getUserId()));
+        }
+    }
+
+    private void shouldThrottle(String userId) throws EmailOtpException {
+
+        String sessionId = Utils.getHash(userId);
+        String jsonString = (String) SessionDataStore.getInstance().
+                getSessionData(sessionId, Constants.SESSION_TYPE_OTP);
+        if (StringUtils.isBlank(jsonString)) {
+            return;
+        }
+
+        SessionDTO previousOTPSessionDTO;
+        try {
+            previousOTPSessionDTO = new ObjectMapper().readValue(jsonString, SessionDTO.class);
+        } catch (IOException e) {
+            throw Utils.handleServerException(Constants.ErrorMessage.SERVER_JSON_SESSION_MAPPER_ERROR, null, e);
+        }
+
+        long elapsedTimeSinceLastOtp = System.currentTimeMillis() - previousOTPSessionDTO.getGeneratedTime();
+        int resendThrottleInterval = EmailOtpServiceDataHolder.getConfigs().getResendThrottleInterval();
+        if (elapsedTimeSinceLastOtp < resendThrottleInterval) {
+            long waitingPeriod = (resendThrottleInterval - elapsedTimeSinceLastOtp) / 1000;
+            throw Utils.handleClientException(
+                    Constants.ErrorMessage.CLIENT_SLOW_DOWN_RESEND, String.valueOf(waitingPeriod));
+        }
+    }
+
+    private int getTenantId() {
+
+        return PrivilegedCarbonContext.getThreadLocalCarbonContext().getTenantId();
```

Comment:
Can't we have a class variable for tenantId and reuse it instead of calling this method? Would improve the performance.

Reply:
In OSGi services in WSO2 normally use singleton instances, so there is no considerable performance hit in calling method level.